theseus/optimizer/nonlinear/nonlinear_least_squares.py

```diff
@@ -32,21 +97,404 @@ def __init__(
+                # context of implicit differentiation. Might be added as a supported
+                # kwarg in the future with a different name, or removed altogether.
+                # If the option is present and set to True, we don't use step size = 1
+                # 1 for the truncated steps, resulting in scaled gradients, but
+                # possibly better solution.
+                if not kwargs.get("__keep_final_step_size__", False):
+                    steps_tensor = torch.ones_like(delta)
+                force_update = True
+            else:
+                force_update = False
+
+            with torch.no_grad():
+                if steps_tensor is None:
+                    steps_tensor = torch.ones_like(delta) * self.params.step_size
+
+            # For now, step size is combined with delta. If we add more sophisticated
+            # line search, will probably need to pass it separately, or compute inside.
+            err, all_rejected = self._step(
+                delta * steps_tensor,
+                info.last_err,
+                converged_indices,
+                force_update,
+                delta_forced_to_gn=_last_implicit_diff_step,
+                **kwargs,
+            )  # err is shape (batch_size,)
+            if all_rejected:
+                # The optimizer rejected all steps so just continue, otherwise convergence
+                # check will stop prematurely. However, we put a max on this to guarantee
+                # this terminates
+                all_reject_attempts += 1
+                if all_reject_attempts < NonlinearOptimizer._MAX_ALL_REJECT_ATTEMPTS:
+                    continue
+            all_reject_attempts = 0
+
+            # check for convergence if at least one step was accepted
+            with torch.no_grad():
+                self._update_info(info, it_, err, converged_indices)
+                if verbose:
+                    print(
+                        f"Nonlinear optimizer. Iteration: {it_+1}. "
+                        f"Error: {err.mean().item()}"
+                    )
+                converged_indices = self._check_convergence(err, info.last_err)
+                info.status[
+                    converged_indices.cpu().numpy()
+                ] = NonlinearOptimizerStatus.CONVERGED
+                if converged_indices.all():
+                    break  # nothing else will happen at this point
+                info.last_err = err
+
+                if end_iter_callback is not None:
+                    end_iter_callback(self, info, delta, it_)
+
+            iters_done += 1
+            it_ += 1
+
+        info.status[
+            info.status == NonlinearOptimizerStatus.START
+        ] = NonlinearOptimizerStatus.MAX_ITERATIONS
+        return iters_done
+
+    # Returns how many iterations to do with and without autograd
+    def _split_backward_iters(self, **kwargs) -> Tuple[int, int]:
```

Comment:
Same goes here, could this be a fn of the NonLinearOptimizer class instead?

Reply:
Moved back up.